lib/mayaUsd/nodes/layerManager.h

```diff
@@ -0,0 +1,117 @@
+//
+// Copyright 2020 Autodesk
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#ifndef MAYA_USD_LAYER_MANAGER
+#define MAYA_USD_LAYER_MANAGER
+
+#include <mayaUsd/base/api.h>
+
+#include <pxr/base/tf/notice.h>
+#include <pxr/pxr.h>
+#include <pxr/usd/sdf/layer.h>
+#include <pxr/usd/sdf/types.h>
+#include <pxr/usd/usd/prim.h>
+
+#include <maya/MMessage.h>
+#include <maya/MObject.h>
+#include <maya/MPxNode.h>
+#include <maya/MString.h>
+#include <maya/MTypeId.h>
+
+#include <functional>
+#include <string>
+
+PXR_NAMESPACE_USING_DIRECTIVE
+
+namespace MAYAUSD_NS_DEF {
+
+typedef std::function<bool(const std::vector<UsdStageRefPtr>&)> BatchSaveDelegate;
+
+/*! \brief Maya dependency node responsible for serializing unsaved Usd edits.
+
+    In a pre-save or export callback, Maya will check if there are any proxy shapes of a supported
+    type that have UsdStages with dirty Layers in them.  If stages are found, there are currently
+    three options for how to handle the Usd edits:
+
+    1. Save back to .usd files.
+    There are three steps to handling this option.
+    a) All anonymous layers must be saved to disk.  If a batch save delegate has been installed,
+    and Maya is running in interactive mode, then a UI dialog can be displayed to provide a choice
+    of file names and locations for all anonymous layers.  If Maya is not running in interactive
+    mode, or there is no installed delegate to handle it, then Maya will automatically choose name
+    and locations for all anonymous layers.
```

Comment:
I think there's a case to be made for a new "Save back to .usd files and keep anonymous layers" case which would save all external files, but not force people to choose file locations for any anonymous layer in the scene. This is a use case we have currently where if a USD file is supposed to be saved on disk we create the file on disk, but for data that we prefer not having to choose a file location then we use anonymous layers. This early decision shouldn't have to be compromised at save time by forcing users to choose between saving their unsaved USD files on disk and giving up on anonymous layers and having to create new files for all.

Reply:
I think this is a good point - both because it makes a lot of sense as default when running in batch mode, as @aloysbaillet points out, and also because I thin it "conceptually fits" with anonymous layers vs file-backed layers.  That is, I think of anonymous layers as more for temporary storage, which might make sense saved out to the maya scene, whereas file-backed layers are generally for more "permanent" changes which you might want to, ie, share with others, publish downstream, etc.

However - my next thought was that, since the session layer is already treated in a special manner (ie, always saved to maya), and the session layer is the only anonymous layer we use at Luma - and likely the only anonymous layer 90%+ of the time - we would effectively already be getting the behavior @aloysbaillet describes, because of the special session-layer handling.

@aloysbaillet - I'm curious, does Animal Logic make a lot of use of non-session-layer anonymous layers? Or were you thinking more of "what if" and "just in case"?

Reply:
Just to be clear, are we talking about putting all anonymous layers into the Maya file (and saving file backed layers to disk) or are you suggesting an option where we ignore anonymous layers entirely (because you consider them kind of temporary)?

Reply:
@aloysbaillet  we talked some internally about your suggestion and I think we’d like to get this merged with the existing options for testing before moving on to additional changes.  Right now the UI does implement what was requested by design so we have completed the task.  I think you also had some questions about overriding some of the file naming/saving behavior, which is what we were starting to think about with the delegate function that the plugin registers to have the UI lib help, so maybe we can work to get this in and then start up a new conversation about next steps.  What do you think?

Reply:
Yes I confirm I would like an option to save all external files, but keep anonymous layers anonymously saved inside the maya file. And I'm happy to wait for that option!
The tension between simplicity and flexibility is quite apparent here, adding another option will make things more confusing for users, and the layer manager can be used to save all external files first, then the option of saving all unsaved layers in the maya scene would work for us.